Fix build on Ubuntu 18.04

anderslanglands opened the pull request:
These changes get the code building for me on stock Ubuntu 18.04 using gcc 7.4. The most invasive part is changing the SphericalTriangle class to contain the A, B and C Pos structs directly rather than in an anonymous union, which is allowed on MSVC and Clang but not GCC. Since the ABC variant of the union doesn't appear to be used anywhere this looks fine to me.

There also appears to be a racy bug where sometimes the isnan assert in `rand_toward_sphericaltri()` is triggered, but I haven't been able to debug that because it's not triggered in gdb for me currently (and adding prints to check the values stops it from occurring).

imallett commented:
Thanks for the contribution; I'll have some comments first, though:

- I don't see any particular reason to allow earlier CMake versions, but it's fine.
- Thanks for pthreads fix.
- Thanks for the `sqrt` fix. However, instead of using `sqrtf`, should use `std::sqrt`, as this is C++.
- Thanks for anonymous unions fix. MSVC is a lot more-flexible than other compilers in this regard—GCC also supports it, but has rules on the types contained which seem bogus to me. But you're correct that this wasn't used, so removing it is fine—and makes the code simpler besides!
- The `+1` is not necessary when seeding the RNG. When `thread_index` is `0` (as it will be on every platform, not just Linux), it still gets hashed, so the resultant RNG state is quite nonzero.
- I'm not able to pin down the `assert` trip on this end. I may be able to look at it more later.

anderslanglands commented:
Personally I think it's better to specify the actual minimum required version of cmake to get your code to build correctly so people don't have to break out of their package manager or add more repos unless it's necessary. Ubuntu 18.04 has 3.10, 16.04 has 3.5 and Centos 7 is all the way back at 2.8 (though I wouldn't recommend going <3.0). It's trivial of course to modify the cmake version oneself, but then you're left wondering if there's something in 3.14 the build was actually using and has the thing built correctly? If you're really not using anything more recent in cmake then I would prefer setting the minimum version to 3.5.

Regarding the thread_index offset - the issue isn't with the thread_id but the fact that std::hasher() returns 0 for a 0 input on gcc and clang (which you can verify here: https://wandbox.org/permlink/Lq8lAXx58EcbGGDY). I assume MSVC does something more sensible.

I'll close this and make a new PR with std::sqrt().

imallett commented:
Re: CMake, for me, I see an out-of-date build tool as a problem with the build environment. (Wherefore also C++17, Ubuntu 19.04, etc.)

Re: `thread_index`; ah, I see the problem now. Yeah, the output of `get_hashed(...)` needs to never be `0`, and there's no particular guarantee imposed by the standard there. `+1` works, but there's no particular reason it should, and I'd prefer something less-arbitrary that fixes the underlying problem. Maybe something like this?

    //Seed the RNG with the thread index (hashed to increase variance).  The seed must be nonzero,
    //	so we enforce that explicitly.
    uint32_t seed = get_hashed(thread_index);
    if (seed==0u) ++seed;
    rng.seed(static_cast<Math::RNG::result_type>(seed));

anderslanglands commented:
Not everyone can change their build environment that easliy :) Lots of production environments in visual effects are locked into stable (i.e. old) versions of linux. The VFX Reference Platform is only just mandating C++14, for example. Anyway, it's not a big deal in this case.

After doing a bit more digging, the problem with the hash is actually a bit worse than I first realized - on GCC and Clang, std::hash of an integral type is the identity function! https://stackoverflow.com/questions/38304877/why-stdhashint-seems-to-be-identity-function
In which case, simply incrementing if the seed is 0 will cause the first two threads to have the same seed. Offsetting might be the best idea after all.

imallett commented:
Oof. Identity is a terrible hash function for avalanching (and that's the main reason for doing this). I had heard of identity being used, but didn't expect it to be in the standard library. But if it's that way on Linux, then I guess one cannot use `std::hash` at all.

What do you think about having the hashing be done "manually"? Thinking some simple public-domain hash specialization for `get_hashed<[Integral]>(...)`. This would allow better consistency between platforms as-well.

Did you want to open a new PR for the other changes (first three commits, except using `std::sqrt`) while we, ah, _hash_ this out? (N.B. I'm a bit new to collaborating on Github, but maybe a separate PR for each would be clearest . . . ?)

anderslanglands commented:
Sure thing I'll make a separate PR for the first chunk. 

For the RNG - I'm not too familiar with PCG, but should't using incrementing seeds be fine?

imallett commented:
Unsure. Most RNGs produce similar or correlated starting numbers for similar seeds (and incrementing clearly produces very similar seeds). I haven't, however, specifically tested if PCG suffers from this problem.

anderslanglands commented:
Maybe Bob Jenkin's hash or a few rounds of TEA to initialize the seed?

imallett commented:
(The rest of the changes [have been merged](https://github.com/imallett/simple-spectral/pull/3); thanks.)

I went ahead and implemented the hashing scheme myself "manually", so it should now work regardless of platform. I did some cursory tests to make sure it works, but I'm currently traveling from the conferences so some kind of sanity check would be welcome.

The solution I went with is basically that of my [previous message](https://github.com/imallett/simple-spectral/pull/2#issuecomment-508970004), except I also implemented the hash directly instead of relying on `std::hash`. Note that the increment from `0` happens after the hashing, and the hashing itself will definitely do some scrambling, so the chance of two threads both being seeded with `1` is low.

anderslanglands commented:
Great! I can confirm a clean checkout, build and run now works from your
repo.

Cheers,
Anders

On Mon, 15 Jul 2019 at 22:11, Ian Mallett <notifications@github.com> wrote:

> (The rest of the changes have been merged
> <https://github.com/imallett/simple-spectral/pull/3>; thanks.)
>
> I went ahead and implemented the hashing scheme myself "manually", so it
> should now work regardless of platform. I did some cursory tests to make
> sure it works, but I'm currently traveling from the conferences so some
> kind of sanity check would be welcome.
>
> The solution I went with is basically that of my previous message
> <https://github.com/imallett/simple-spectral/pull/2#issuecomment-508970004>,
> except I also implemented the hash directly instead of relying on
> std::hash. Note that the increment from 0 happens after the hashing, and
> the hashing itself will definitely do some scrambling, so the chance of two
> threads both being seeded with 1 is low.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/imallett/simple-spectral/pull/2?email_source=notifications&email_token=AAOYQXK2GAZ5537H6VP2WFLP7REN7A5CNFSM4H6IITGKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZ5IGPI#issuecomment-511345469>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAOYQXP5MUUD5CKJWYPE6RTP7REN7ANCNFSM4H6IITGA>
> .
>
